### PR TITLE
Allow hook to inspect leader key sequence as it grows, and provide example code to exit the leader mode early if a prefix is matched

### PR DESCRIPTION
**Pull request by tennysontbardwell**
## Description

This is similar to other efforts to implement a compose key (e.g., https://github.com/qmk/qmk_firmware/pull/10063/files) but more modular, allowing for fewer code changes and more customizability for the user.

Aside from the features covered in the title, this also fixes a bug. See line 76 in `process_leader.c`. When `LEADER_NO_TIMEOUT` is defined and line `prcoess_leader.c:55` is run, `leading` will remain `true`, breaking the leading mode until a reset occurs.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* `LEADER_NO_TIMEOUT` could break the leader mode

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by tennysontbardwell** [comment]
Thank you @zvecr  for fixing the branch

File: docs/feature_leader_key.md [hunk 1]
```diff
@@ -41,6 +41,46 @@ As you can see, you have a few functions. You can use `SEQ_ONE_KEY` for single-k
 
 Each of these accepts one or more keycodes as arguments. This is an important point: You can use keycodes from **any layer on your keyboard**. That layer would need to be active for the leader macro to fire, obviously.
 
+## Leader_new_key
+Alternatively, the leader key also offers a hook to inspect the leader sequence as it grows. This allows for short circuiting when a prefix is matched.
```
Comment (spidey3) [inline docs/feature_leader_key.md:45]: Please expand this paragraph to provide more detail about `leader_new_key()`:

- When is it called?
- How it it envisioned to be used?
- How does the included example use it?

File: quantum/process_keycode/process_leader.c [hunk 2]
```diff
@@ -70,6 +73,11 @@ bool process_leader(uint16_t keycode, keyrecord_t *record) {
 #    endif
                 return false;
             }
+#    ifndef LEADER_NO_TIMEOUT
```
Comment (spidey3) [inline quantum/process_keycode/process_leader.c:76]: This change seems unrelated to the topic of the PR. 
Can you please either remove it from this PR, or else add an explanation of this part of the change to the PR description?

**Review by spidey3 — CHANGES_REQUESTED** [review]
Thanks for your submission; this looks useful!
I do have a few comments - mainly about documentation (see above).

File: quantum/process_keycode/process_leader.c (hunk 2 above)
Comment (tennysontbardwell) [reply quantum/process_keycode/process_leader.c:76]: From my description:

> Aside from the features covered in the title, this also fixes a bug. See line 76 in process_leader.c. When LEADER_NO_TIMEOUT is defined and line prcoess_leader.c:55 is run, leading will remain true, breaking the leading mode until a reset occurs.

Happy to promote to its own PR, drop the change, or leave as is.

File: docs/feature_leader_key.md (hunk 1 above)
Comment (tennysontbardwell) [reply docs/feature_leader_key.md:45]: How about something like:

> The function `leader_new_key` is called on every keypress during leading mode. To see the sequence of keypresses, inspect the first `leader_sequence_size` elements of `leader_sequence`. If you wish to exit leader mode for any reason, setting `leading` to `false` inside this function will accomplish that end.

> For example, the following will check `leader_sequence` against a set of macros as the user types, and will detect, run, and set `leading = false` immediately upon the final keypress:

File: docs/feature_leader_key.md [hunk 3]
```diff
@@ -41,6 +41,46 @@ As you can see, you have a few functions. You can use `SEQ_ONE_KEY` for single-k
...
+    } else if (leader_match( (uint16_t[]) { KC_P, KC_P }, 2)) { // personal phone
+        SEND_STRING("888-888-8888");
+        leading = false;
+    }
+}
+
+void leader_start(void) {}
+
+void leader_end(void) {
+    if (did_leader_succeed) {
```
Comment (tennysontbardwell) [inline docs/feature_leader_key.md:78]: TODO: remove reference to `did_leader_succeed`

File: docs/feature_leader_key.md (hunk 1 above)
Comment (spidey3) [reply docs/feature_leader_key.md:45]: That's better!

File: quantum/process_keycode/process_leader.c (hunk 2 above)
Comment (spidey3) [reply quantum/process_keycode/process_leader.c:76]: I think it can remain here.

File: docs/feature_leader_key.md (hunk 3 above)
Comment (spidey3) [reply docs/feature_leader_key.md:78]: Can you elaborate on this?

Comment (spidey3) [reply docs/feature_leader_key.md:78]: @tennysontbardwell - I'd like to clear this review off of my plate. What exactly did you mean by:
> TODO: remove reference to `did_leader_succeed`
